### PR TITLE
RavenDB-16955: add traffic watch logging tool

### PR DESCRIPTION
**Pull request by garayx**

### CODE REVIEW

File: tools/rvn/LogTrafficWatch.cs [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+            _path = path;
+            _changeTypes = changeTypes;
+            if (string.IsNullOrEmpty(certPath))
+            {
+                _url = url.StartsWith("http://") == false ? $"http://{url}" : url;
+            }
+            else
+            {
+                _cert = new X509Certificate2(certPath);
+                _url = url.StartsWith("https://") == false ? $"https://{url}" : url;
```
Comment (ayende) [inline tools/rvn/LogTrafficWatch.cs]: Use `UriBuilder`, maybe? Or at least verify that we aren't going to get: `https://http://foo`

File: tools/rvn/LogTrafficWatch.cs [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+                                        var t = Task.Run(() =>
+                                        {
+                                            using (var logStream = SafeFileStream.Create(oldFileName, FileMode.Open, FileAccess.Read))
+                                            using (var newFileStream = SafeFileStream.Create(oldFileName + ".gz", FileMode.Create, FileAccess.Write))
+                                            using (var compressionStream = new GZipStream(newFileStream, CompressionMode.Compress))
+                                            {
+                                                logStream.CopyTo(compressionStream);
+                                            }
+
+                                            File.Delete(oldFileName);
```
Comment (ayende) [inline tools/rvn/LogTrafficWatch.cs]: There might be a race here with closing of the old file stream, mind.

**Review by ayende — CHANGES_REQUESTED** [review]
I really like it, I had some more notes along the way, and you address all of them

File: tools/rvn/LogTrafficWatch.cs (hunk 2 above)
Comment (garayx) [reply tools/rvn/LogTrafficWatch.cs]: done

File: tools/rvn/CommandLineApp.cs [hunk 3]
```diff
@@ -89,6 +92,94 @@ private static void ConfigureLogsCommand()
...
 
+        private static void ConfigureTrafficWatchCommand()
+        {
+            _app.Command("traffic-watch", cmd =>
+            {
+                cmd.ExtendedHelpText = cmd.Description = "Log traffic watch entries.";
+                cmd.HelpOption(HelpOptionString);
+
+                var urlArg = cmd.Argument("Url", "Server URL.");
+                var pathArg = cmd.Argument("Path", "Path to save the traffic watch log.");
```
Comment (ppekrol) [inline tools/rvn/CommandLineApp.cs:103]: OutputPath

File: tools/rvn/CommandLineApp.cs [hunk 4]
```diff
@@ -89,6 +92,94 @@ private static void ConfigureLogsCommand()
...
+                var typesArg = cmd.Argument("Types", "Path to save the traffic watch log.");
```
Comment (ppekrol) [inline tools/rvn/CommandLineApp.cs:104]: Wrong description

File: tools/rvn/CommandLineApp.cs [hunk 5]
```diff
@@ -89,6 +92,94 @@ private static void ConfigureLogsCommand()
...
+                var certArg = cmd.Argument("Certificate", "Path to certificate.");
```
Comment (ppekrol) [inline tools/rvn/CommandLineApp.cs:105]: CertificatePath

File: tools/rvn/LogTrafficWatch.cs [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+        public LogTrafficWatch(int? pid, string url, string certPath, string path, List<TrafficWatchChangeType>? changeTypes)
+        {
+            _pid = pid ?? ServerProcessUtil.GetRavenServerPid();
+            _path = path;
+            _changeTypes = changeTypes;
+            if (string.IsNullOrEmpty(certPath))
+            {
+                if (url.StartsWith("https://"))
+                {
+                    ExitAndPrintError(ex: null, Environment.NewLine + $"URL '{url}' starts with 'https' but no certificate provided.");
```
Comment (ppekrol) [inline tools/rvn/LogTrafficWatch.cs:44]: `no certificate was provided`

File: tools/rvn/LogTrafficWatch.cs [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+                }
+                _url = url.StartsWith("http://") == false ? $"http://{url}" : url;
+            }
+            else
+            {
+                if (url.StartsWith("http://"))
+                {
+                    ExitAndPrintError(ex: null, Environment.NewLine + $"URL '{url}' starts with 'http' but certificate provided.");
```
Comment (ppekrol) [inline tools/rvn/LogTrafficWatch.cs:52]: `but certificate was provided`

File: tools/rvn/LogTrafficWatch.cs [hunk 8]
```diff
@@ -0,0 +1,237 @@
...
+                }
+                _cert = new X509Certificate2(certPath);
```
Comment (ppekrol) [inline tools/rvn/LogTrafficWatch.cs:54]: we need to dispose this?

File: tools/rvn/LogTrafficWatch.cs [hunk 9]
```diff
@@ -0,0 +1,237 @@
...
+                try
+                {
+                    await _client.ConnectAsync(url, _cts.Token).ConfigureAwait(false);
+                }
+                catch (Exception ex)
+                {
+                    ExitAndPrintError(ex, Environment.NewLine + "Couldn't connect to " + _url);
+                }
+
+                Console.WriteLine($"Connected to RavenDB server. Collecting traffic watch entries to {_path}");
```
Comment (ppekrol) [inline tools/rvn/LogTrafficWatch.cs:96]: wrap path with `''`

File: tools/rvn/LogTrafficWatch.cs [hunk 10]
```diff
@@ -0,0 +1,237 @@
...
+                                            continue;
+                                        }
+                                    }
+
+                                    // log to file
+                                    json.WriteJsonTo(fileStream);
+                                    await fileStream.FlushAsync();
+                                }
+
+                                if (new FileInfo(file).Length >= maxFileSize)
```
Comment (ppekrol) [inline tools/rvn/LogTrafficWatch.cs:152]: there is a method called `Refresh` in FileInfo

**Pull request closed without merging**